rasa_nlu_examples/featurizers/sparse/hashing_featurizer.py

```diff
@@ -0,0 +1,139 @@
+    https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.HashingVectorizer.html
+    """
+
+    @classmethod
+    def required_components(cls) -> List[Type[Component]]:
+        return [Tokenizer]
+
+    @classmethod
+    def required_packages(cls) -> List[Text]:
+        return ["sklearn"]
+
+    defaults = {
+        # The following parameters and defaults are similar as the ones used by
+        # scikit-learn (version 0.24.2). For detailed explanations on these parameters
+        # and their defaults, have a look at the scikit-learn docs:
+        # https://scikit-learn.org/0.24/modules/generated/sklearn.feature_extraction.text.HashingVectorizer.html
+        #
+        # the number of columns in the output vector
+        "n_features": 2 ** 20,  # int
+        # whether to use word or character n-grams
+        # 'char_wb' creates character n-grams inside word boundaries
+        # n-grams at the edges of words are padded with space.
+        "analyzer": "word",  # use 'char' or 'char_wb' for character
+        # remove accents during the preprocessing step
+        "strip_accents": None,  # {'ascii', 'unicode', None}
+        # list of stop words
+        "stop_words": None,  # string {'english'}, list, or None (default)
+        # if `True` convert all characters to lowercase
+        "lowercase": True,  # bool
+        # set range of ngrams to be extracted
+        "ngram_range": (1, 1),
+        # if `True`, all non zero counts are set to 1.
+        "binary": False,  # bool
+        # the norm used to normalize term vectors
+        "norm": "l2",  # 'l1', 'l2', or None
+        # when `True`, an alternating sign is added to the features as to approximately
+        # conserve the inner product in the hashed space even for small n_features.
+        "alternate_sign": True,  # bool
+    }
+
+    def __init__(self, component_config: Optional[Dict[Text, Any]] = None) -> None:
+        super().__init__(component_config)
+        self.hashing_vectorizer = HashingVectorizer(
+            **{key: self.component_config[key] for key in self.defaults},
+            token_pattern=r"(?u)\b\w+\b"
+            if self.component_config["analyzer"] == "word"
+            else None,
+            dtype=np.float32,
+        )
+
+    def train(
+        self,
+        training_data: TrainingData,
+        config: Optional[RasaNLUModelConfig] = None,
+        **kwargs: Any,
+    ) -> None:
+        for example in training_data.intent_examples:  # type: ignore
+            for attribute in DENSE_FEATURIZABLE_ATTRIBUTES:
+                self.set_features(example, attribute)
+
+    def create_word_vector(self, document: List[Text]) -> np.ndarray:
+        return self.hashing_vectorizer.fit_transform(document).tocoo()
+
+    def set_features(self, message: Message, attribute: Text = TEXT) -> None:
+        tokens = message.get(TOKENS_NAMES[attribute])
+
+        if not tokens:
+            return None
```

Comment:
As mentioned in the unit test, I'm not 100% sure if this is good behavior. Won't DIET break without any features? Any empty array may be better.

Reply:
This was mimicked from the behaviour of the other feature extractors in this repo. I checked in Rasa's count vectorizer and [here](https://github.com/RasaHQ/rasa/blob/main/rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py#L551) it seems that if no tokens are present, `None` is added to the list of features, which in turn leads to no features being added by [this function](https://github.com/RasaHQ/rasa/blob/main/rasa/nlu/featurizers/featurizer.py#L61).

A quick test with moodbot and NLU config
```
- name: WhitespaceTokenizer
- name: rasa_nlu_examples.featurizers.sparse.HashingFeaturizer
  n_features: 2048
- name: DIETClassifier
  epochs: 100
- name: FallbackClassifier
  threshold: 0.3
  ambiguity_threshold: 0.1
```
lead to 
```
2021-11-29 13:30:28 DEBUG    rasa.core.processor  - Received user message '' with intent '{'name': None, 'confidence': 0.0}' and entities '[]'
```
for empty user input. So DIET doesn't seem to break, but I'm not sure if this would be an issue for the dialogue manager later on? I can test with an empty array instead, how that changes the behaviour. In that case the other featurizers could be adapted as well.

Reply:
@ka-bu, have you gotten a strong opinion about this? If we mimic the countvectorizer behavior I suppose I'm fine, but my dense vectorizers would emit zeros in this case.

Reply:
Don't the fasttext, gensim and bpemb featurizers follow a similar pattern in this case, i.e. not setting features? E.g. [here](https://github.com/RasaHQ/rasa-nlu-examples/blob/main/rasa_nlu_examples/featurizers/dense/bpemb_featurizer.py#L406)

Reply:
I stand corrected! I seem to be confusing the whatlies implementation with the Rasa one. In order for whatlies to be scikit-learn compatible, we demand that they return a numeric vector on an empty string. 

My bad!

Reply:
Ah, great, thanks! Then I'd omit the empty array test.

Reply:
I'm wondering whether we catch that beforehand or whether the tensor conversion later fills in a "fake feature" later in case an attribute is missing completely 🤔 . @mleimeister could you inspect the features produced for DIET please? (e.g. place a breakpoint at the end of the` _create_model_data` function of DIET (if `training` == False) and inspect the rasa model data for the empty message)